src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java

```diff
@@ -0,0 +1,159 @@
+                return getResource(name, check);
+            }
+            return null;
+        }
+        return delegate.getResource(name, check);
+    }
+
+    @Override
+    public URL findResource(final String name, boolean check) {
+        URLClassPath delegate = index.get(name);
+        if (delegate == null) {
+            if (lastIndexed < urls.length) {
+                indexURLs(urls[lastIndexed]);
+                lastIndexed ++;
+                return findResource(name, check);
+            }
+            return null;
+        }
+        return delegate.findResource(name, check);
+    }
+
+
+    public List<String> listPackageContents(final String packageName, boolean check) {
+
+        String packageFile = packageName.replaceAll("\\.", File.separator);
+        while (lastIndexed < urls.length) {
+            indexURLs(urls[lastIndexed]);
+            lastIndexed ++;
+        }
+        return index.list(packageFile);
+    }
+
+    public List<String> listSDKContents(final String packageName, boolean check) {
+        String packageFile = packageName.replaceAll("\\.", File.separator);
+        indexURLs(RT_JAR);
+        return index.list(packageFile);
+    }
+
+
+    private void indexURLs(URL url) {
+       try {
+           if (!"file".equals(url.getProtocol())) {
+               throw new RuntimeException("Classpath element is not a file: " + url);
+           }
+           File root = new File(url.getPath());
+           URL[] args = {url};
+           URLClassPath delegate = new URLClassPath(args);
+
+           if (root.isDirectory()) {
+               String rootPath = root.getPath();
+               if (!rootPath.endsWith(File.separator)) {
+                   rootPath = rootPath + File.separator;
+               }
+               addFilesToIndex(rootPath.length(), root, delegate);
+           } else if (root.isFile() && root.getName().endsWith(".jar")) {
+               JarFile jarFile = new JarFile(root);
+               try {
+                   Enumeration<JarEntry> entries = jarFile.entries();
+                   while (entries.hasMoreElements()) {
+                       JarEntry entry = entries.nextElement();
+                       String name = entry.getName();
+                       maybeIndexResource(name, delegate);
+                   }
+               } finally {
+                   jarFile.close();
+               }
+           }
+       } catch (IOException e) {
+           throw new RuntimeException(e);
+       }
+    }
+
+    private void addFilesToIndex(int basePrefixLen, File f, URLClassPath delegate) throws IOException {
+        if (f.isDirectory()) {
+            if (f.getPath().length() > basePrefixLen) {  // Don't index the root itself.
+                String relPath = f.getPath().substring(basePrefixLen);
+                maybeIndexResource(relPath, delegate);
+            }
+            File[] directoryEntries = f.listFiles();
+            assert(directoryEntries != null);
```

Comment:
Throw ioexception instead of assert.

Reply:
it was because the code was initially copied from the pants project

Reply:
a95af1e0961d628eb62289eacc0b2d2c32b6272b